### ISSUE TITLE
Detect File Disclosure Injection with Spring's RequestMapping and prefixed redirect string

### ISSUE THREAD

**Issue by ptamarit** [find-sec-bugs/find-sec-bugs#269]
In the PR #265, checks for File Disclosure Injection has been added for Spring when using `ModelAndView`.

Redirects are also possible in Spring by return a String with the `"redirect:"` prefix within a method annotated with `@RequestMapping`, `@GetMapping`, etc.

Vulnerable code snippet:

```java
@GetMapping
public String index(...) {
    // ...
    return "redirect:" + request.getParameter("returnURL");
}
```

References:
* https://docs.spring.io/spring/docs/current/spring-framework-reference/html/mvc.html#mvc-redirecting-redirect-prefix
* http://stackoverflow.com/questions/19516719/difference-between-redirect-and-redirect-inside-modelandview

I'm afraid I don't have enough FindBugs knowledge to implement this one...

**Comment by h3xstream** [comment]
Note : It will be much easier to implement once the new Taint Analysis extensions are implemented. 
https://github.com/find-sec-bugs/find-sec-bugs/pull/220#issuecomment-274403955

**Comment by php-coder** [comment]
@ptamarit This example is not vulnerable to File Disclosure Injection. Original issue was about server-side redirect. This example is using client side redirects and the most dangerous thing that it could do, is to redirect user to some malicious site (see https://www.owasp.org/index.php/Unvalidated_Redirects_and_Forwards_Cheat_Sheet#Dangerous_URL_Redirects).

To be vulnerable to server-side redirects, controller should use "forward:" instead of "redirect:". See https://docs.spring.io/spring/docs/current/spring-framework-reference/html/mvc.html#mvc-redirecting-forward-prefix

**Comment by h3xstream** [comment]
I am doing some test right now to verify the extend. The precision @php-coder seems logical. It should only be URL.

**Comment by php-coder** [comment]
@h3xstream Let me know I you need my help (in case you're not familiar with Spring/Spring Boot).

**Comment by h3xstream** [comment]
Yes it is only a URL redirects (client-side) if the `redirect:` prefix is used.
By the way, the redirection does not filter any protocol : `javascript:`, `data:`, `file:///` or even application.

I stumble upon an additional vulnerability...

**Comment by h3xstream** [comment]
Here is an additional vulnerability that is not cause by concatenation..

```java
    @RequestMapping("/redirect")
    public ModelAndView redirect(ModelAndView view) {

        view.addObject("userInfo", getUserFromDatabase());
        view.addObject("secret","test1234");

        if(valid) {
            view.setViewName("index");
        }
        else {
            view.setViewName("redirect:/login");
        }
        return view;
    }
```

The behavior is probably not was is expected by the developer. Here, the redirect location will be follow by (`?userInfo=[User...]&secret=test1234`)

It is also possible to use bracket (`{secret}`) in the url return to exposed a specific object of the model. This is not particularly malicious when the default behavior append all the model.

**Comment by h3xstream** [comment]
# Detector proposed

There is already a detector propose that look for concatenation of "redirect:" and a dynamic value.

https://github.com/johnhawes/find-sec-bugs/blob/23bdae2a0a7aa7cf188c110e89fc0c93b2481cb3/plugin/src/main/resources/metadata/messages.xml#L3173

The description seems ok. It is a specialized version of the detector `UnvalidatedRedirectDetector`

**Comment by h3xstream** [comment]
@ptamarit Let me know if some vulnerable cases still need to be cover.
I close the issue after.

**Comment by php-coder** [comment]
What's about covering the case of using "forward:" prefix?

**Comment by h3xstream** [comment]
forward is only doing redirection to external URL to the same website. I don't think it can give an attacker access to more content.

**Comment by php-coder** [comment]
> "forward:" is only doing redirection to external URL to the same website. I don't think it can give an attacker access to more content.

Actually it can. And that's why `ActionForward` has `Forward` in its name and this is what this class is doing:

> An ActionForward represents a destination to which the controller servlet, ActionServlet, might be directed to perform a RequestDispatcher.forward()

See: https://svn.apache.org/repos/asf/struts/archive/trunk/struts-doc-1.1/api/org/apache/struts/action/ActionForward.html

@h3xstream If the check from #265 don't protect from forwards, then I don't understand from what it tries to protect?

**Comment by h3xstream** [comment]
@php-coder Yes I'm totally about covering `forward:`.
I just don't want it to flagging unvalidated redirect (which is intented to be client-side).

It should be raising the same bug type as "FileDisclosureDetector". And as you point out, the description need to be updated to reflect the actual risk of controlling the view name.

**Comment by ptamarit** [comment]
@h3xstream You did much more research than me on the topic. If you think that everything is covered, please go ahead and close the ticket.
